x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java

```diff
@@ -238,34 +247,73 @@ protected TermsEnumResponse newResponse(
+                    shardFailures.add(new DefaultShardOperationFailedException(remote.clusterAlias + ":" + exc.index(),
+                        exc.shardId(), exc.getCause()));
+                }
+                List<TermCount> terms = remote.resp.getTerms().stream()
+                    .map(a -> new TermCount(a, 1))
+                    .collect(Collectors.toList());
+                nodesTerms.add(terms);
+            } else {
+                throw new AssertionError("Unknown node response type: " + nodeResponse.getClass().getName());
             }
         }
-        int size = Math.min(request.size(), combinedResults.size());
-        List<String> terms = new ArrayList<>(size);
-        TermCount[] sortedCombinedResults = combinedResults.values().toArray(new TermCount[0]);
-        // Sort alphabetically
-        Arrays.sort(sortedCombinedResults, new Comparator<TermCount>() {
-            public int compare(TermCount t1, TermCount t2) {
-                return t1.getTerm().compareTo(t2.getTerm());
+
+        final PriorityQueue<TermCountIterator> pq = new PriorityQueue<>(nodesTerms.size()) {
+            @Override
+            protected boolean lessThan(TermCountIterator a, TermCountIterator b) {
+                return a.compareTo(b) < 0;
             }
-        });
+        };
 
-        for (TermCount term : sortedCombinedResults) {
-            terms.add(term.getTerm());
-            if (terms.size() == size) {
-                break;
+        for (List<TermCount> nodeTerms : nodesTerms) {
+            Iterator<TermCount> it = nodeTerms.iterator();
+            if (it.hasNext()) {
+                pq.add(new TermCountIterator(it));
             }
         }
-        return new TermsEnumResponse(terms, (failedShards + successfulShards), successfulShards, failedShards, shardFailures, complete);
+
+        TermCount lastTerm = null;
+        final List<String> results = new ArrayList<>();
+        while (pq.size() != 0) {
+            TermCountIterator it = pq.top();
+            String term = it.term();
+            long docCount = it.docCount();
+            if (lastTerm != null && lastTerm.getTerm().compareTo(term) != 0) {
+                results.add(lastTerm.getTerm());
+                if (results.size() == request.size()) {
+                    break;
+                }
+                lastTerm = null;
+            }
+            if (lastTerm == null) {
+                lastTerm = new TermCount(term, 0);
+            }
+            lastTerm.addToDocCount(docCount);
+            if (it.hasNext()) {
+                String itTerm = it.term();
+                it.next();
+                assert itTerm.compareTo(it.term()) <= 0;
+                pq.updateTop();
+            } else {
+                pq.pop();
+            }
+        }
+        if (lastTerm != null && results.size() < request.size()) {
+            results.add(lastTerm.getTerm());
+        }
+
```

Comment:
I ended up rewriting the logic to do a merge sort with a priority queue. The change is small but I can open a dedicated PR if that's preferred.